cdc/sinkv2/ddlsink/mq/mq_ddl_sink.go

```diff
@@ -0,0 +1,166 @@
+// limitations under the License.
+
+package mq
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/contextutil"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sink/mq/codec"
+	"github.com/pingcap/tiflow/cdc/sink/mq/dispatcher"
+	"github.com/pingcap/tiflow/cdc/sink/mq/manager"
+	"github.com/pingcap/tiflow/cdc/sinkv2/ddlsink"
+	"github.com/pingcap/tiflow/cdc/sinkv2/ddlsink/mq/producer"
+	"github.com/pingcap/tiflow/pkg/config"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/util"
+	"go.uber.org/zap"
+)
+
+// Assert DDLEventSink implementation
+var _ ddlsink.DDLEventSink = (*ddlSink)(nil)
+
+type ddlSink struct {
+	// id indicates which processor (changefeed) this sink belongs to.
+	id model.ChangeFeedID
+	// role indicates what this sink is used for.
+	role util.Role
+	// protocol indicates the protocol used by this sink.
+	protocol config.Protocol
+	// eventRouter used to route events to the right topic and partition.
+	eventRouter *dispatcher.EventRouter
+	// topicManager used to manage topics.
+	// It is also responsible for creating topics.
+	topicManager manager.TopicManager
+	// encoderBuilder builds encoder for the sink.
+	encoderBuilder codec.EncoderBuilder
+	// producer used to send events to the MQ system.
+	// Usually it is a sync producer.
+	producer producer.Producer
+}
+
+func newDDLSink(ctx context.Context,
+	producer producer.Producer,
+	topicManager manager.TopicManager,
+	eventRouter *dispatcher.EventRouter,
+	encoderConfig *codec.Config,
+) (*ddlSink, error) {
+	changefeedID := contextutil.ChangefeedIDFromCtx(ctx)
+	role := contextutil.RoleFromCtx(ctx)
+
+	encoderBuilder, err := codec.NewEventBatchEncoderBuilder(ctx, encoderConfig)
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrKafkaInvalidConfig, err)
+	}
+
+	s := &ddlSink{
+		id:             changefeedID,
+		role:           role,
+		protocol:       encoderConfig.Protocol(),
+		eventRouter:    eventRouter,
+		topicManager:   topicManager,
+		encoderBuilder: encoderBuilder,
+		producer:       producer,
+	}
+
+	return s, nil
+}
+
+func (k *ddlSink) WriteDDLEvent(ctx context.Context, ddl *model.DDLEvent) error {
+	encoder := k.encoderBuilder.Build()
+	msg, err := encoder.EncodeDDLEvent(ddl)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if msg == nil {
+		return nil
+	}
+
+	topic := k.eventRouter.GetTopicForDDL(ddl)
+	partitionRule := k.eventRouter.GetDLLDispatchRuleByProtocol(k.protocol)
+	log.Debug("emit ddl event",
+		zap.Uint64("commitTs", ddl.CommitTs),
+		zap.String("query", ddl.Query),
+		zap.String("namespace", k.id.Namespace),
+		zap.String("changefeed", k.id.ID),
+		zap.Any("role", k.role))
+	if partitionRule == dispatcher.PartitionAll {
+		partitionNum, err := k.topicManager.GetPartitionNum(topic)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = k.producer.SyncBroadcastMessage(ctx, topic, partitionNum, msg)
```

Comment:
Is there a time limit for this operation?

Reply:
No, it will be a sync write.

Reply:
We write DDL asynchronously in Owner, so I think it's OK to be time-consuming here.

Reply:
I think we should add a log if the duration of the execution of `SyncBroadcastMessage` it too long. It helps diagnosing problems online.